File: julia/Algorithms/test/test_trees.jl
const MAX_INT = 10
const COUNT = 10
const REPEAT = 40

@testset "trees" begin
    @testset "test_binary" begin
        data = Set()
        tree = Algorithms.BinaryTree()
        for c = 1 : COUNT
            for i = 1 : c
                value = rand(-MAX_INT:MAX_INT)
                Algorithms.insert(tree, value)
                push!(data, value)
                @test Algorithms.inorder(tree) == sort(collect(data))
            end
            for value in data
                Algorithms.delete(tree, value)
                pop!(data, value)
                @test Algorithms.inorder(tree) == sort(collect(data))
            end
            @test Algorithms.count(tree) == 0
            @test Algorithms.inorder(tree) == []
        end
    end

    @testset "test_binary1" begin
        tree = Algorithms.BinaryTree()
        Algorithms.insert(tree, 50)
        Algorithms.insert(tree, 30)
        Algorithms.insert(tree, 20)
        Algorithms.insert(tree, 40)
        Algorithms.insert(tree, 70)
        Algorithms.insert(tree, 60)
        Algorithms.insert(tree, 80)
        @test Algorithms.inorder(tree) == [20, 30, 40, 50, 60, 70, 80]
        Algorithms.delete(tree, 20)
        @test Algorithms.inorder(tree) == [30, 40, 50, 60, 70, 80]
        Algorithms.delete(tree, 30)
        @test Algorithms.inorder(tree) == [40, 50, 60, 70, 80]
        Algorithms.delete(tree, 50)
        @test Algorithms.inorder(tree) == [40, 60, 70, 80]
        Algorithms.delete(tree, 80)
        @test Algorithms.inorder(tree) == [40, 60, 70]
    end
end

File: python/tests/test_trees.py
# -*- coding: utf-8 -*-
import trees
import random

MAX_INT = 10
COUNT = 10
REPEAT = 40


def test_binary():
    data = set()
    tree = trees.Tree()
    for c in range(1, COUNT + 1):
        for i in range(c):
            value = random.randint(-MAX_INT, MAX_INT)
            tree.insert(value)
            data.add(value)
            assert list(tree) == sorted(list(data))
        for value in list(data):
            tree.delete(value)
            data.remove(value)
            assert list(tree) == sorted(list(data))
        assert list(tree) == []


def test_binary1():
    tree = trees.Tree()
    tree.insert(50)
    tree.insert(30)
    tree.insert(20)
    tree.insert(40)
    tree.insert(70)
    tree.insert(60)
    tree.insert(80)
    assert list(tree) == [20, 30, 40, 50, 60, 70, 80]
    tree.delete(20)
    assert list(tree) == [30, 40, 50, 60, 70, 80]
    tree.delete(30)
    assert list(tree) == [40, 50, 60, 70, 80]
    tree.delete(50)
    assert list(tree) == [40, 60, 70, 80]
